runner/app/utils/hardware.py

```diff
@@ -1,77 +1,114 @@
+"""Contains utility functions for hardware information."""
 
-from typing import Dict
+from typing import Dict, Any
 from pydantic import BaseModel
-import torch
-import pynvml
 import logging
+import pynvml
+from app.utils.nvml_manager import nvml_manager
 
 logger = logging.getLogger(__name__)
 
-class HardwareDetail(BaseModel):
+
+class BaseHardwareDetail(BaseModel):
+    """Base model for GPU information."""
+
     id: str
     name: str
     memory_total: int
     memory_free: int
+
+
+class HardwareDetail(BaseHardwareDetail):
+    """Model for GPU information."""
+
     major: int
     minor: int
 
-class HardwareStatDetail(BaseModel):
-    id: str
-    name: str
-    memory_free: int
-    memory_total: int
+
+class HardwareStatDetail(BaseHardwareDetail):
+    """Model for real-time GPU statistics."""
+
     utilization_compute: int
     utilization_memory: int
 
-def get_cuda_devices() -> Dict[int, HardwareDetail]:
+
+def retrieve_cuda_info(
+    cuda_version: bool = False, utilization: bool = False
+) -> Dict[int, Dict[str, Any]]:
+    """Retrieve CUDA information.
+
+    Args:
+        cuda_version: Whether to retrieve CUDA version information.
+        utilization: Whether to retrieve GPU utilization information.
+
+    Returns:
+        CUDA information.
+    """
+    nvml_manager.initialize()
     devices = {}
-    try:
-        pynvml.nvmlInit()
-        for i in range(pynvml.nvmlDeviceGetCount()):
-            handle = pynvml.nvmlDeviceGetHandleByIndex(i)
-            uuid = pynvml.nvmlDeviceGetUUID(handle)
-            name = pynvml.nvmlDeviceGetName(handle)
+    for i in range(pynvml.nvmlDeviceGetCount()):
+        handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+        uuid = pynvml.nvmlDeviceGetUUID(handle)
+        name = pynvml.nvmlDeviceGetName(handle)
+        memory_info = pynvml.nvmlDeviceGetMemoryInfo(handle)
+
+        device_info = {
+            "id": uuid,
+            "name": name,
+            "memory_total": memory_info.total,
+            "memory_free": memory_info.free,
+        }
+
+        if cuda_version:
             major, minor = pynvml.nvmlDeviceGetCudaComputeCapability(handle)
-            memory_info = pynvml.nvmlDeviceGetMemoryInfo(handle)
-
-            devices[i] = HardwareDetail(
-                id=uuid,
-                name=name,
-                memory_total=memory_info.total,
-                memory_free=memory_info.free,
-                major=major,
-                minor=minor
-            )
-    except BaseException as e:
-        logger.info(f"Error getting cuda devices: error={e}")
-    finally:
-        pynvml.nvmlShutdown()
-    
+            device_info["major"] = major
```

Comment:
Why switch back to manually setting the keys when we have the classes?

Reply:
You have a good point. How about using https://github.com/ad-astra-video/ai-worker/pull/1/commits/511f08f26bd73c2a49ed1c6837d6a0058674a13b.